apps/passport-client/components/modals/AccountImportModal.tsx

```diff
@@ -0,0 +1,253 @@
+
+    dispatch({
+      type: "merge-import",
+      collection: importState.collection,
+      pcdsToMergeIds: importState.pcdsToMergeIds
+    });
+
+    setImportState({ state: "import-complete" });
+  }, [dispatch, importState]);
+
+  // Responds to the user having selected a file to import, or to changes in
+  // the user's current PCD collection.
+  useEffect(() => {
+    (async () => {
+      // If a file has been selected, and isn't invalid or already imported
+      if (
+        filesContent.length > 0 &&
+        importState.state !== "import-complete" &&
+        importState.state !== "invalid-file"
+      ) {
+        let importedCollection: PCDCollection;
+
+        // If the file hasn't been processed yet, process it
+        if (importState.state === "ready") {
+          try {
+            // Parse the file content as JSON
+            const storageExport = JSON.parse(filesContent[0].content);
+            // Deserialize the storage - throws an error if the content is not
+            // recognized
+            const importedBackup = await deserializeStorage(
+              storageExport,
+              await getPackages()
+            );
+            importedCollection = importedBackup.pcds;
+            if (!(importedCollection instanceof PCDCollection)) {
+              throw new Error("Did not deserialize a valid PCD collection");
+            }
+          } catch (e) {
+            // The file is not valid, so bail out
+            setImportState({ state: "invalid-file" });
+            return;
+          }
+        } else {
+          importedCollection = importState.collection;
+        }
+
+        const userHasSemaphoreIdentity =
+          pcdCollection.getPCDsByType(SemaphoreGroupPCDTypeName).length > 0;
+
+        const userHasEmailPCD =
+          pcdCollection.getPCDsByType(EmailPCDTypeName).length > 0;
+
+        // Before importing, we want to filter the PCDs down to those which
+        // are valid to import, so we can tell the user how many new PCDs to
+        // expect
+        const preImportFilter = (pcd: PCD) => {
+          // If the user has a semaphore identity PCD, don't import another
+          if (
+            userHasSemaphoreIdentity &&
+            pcd.type === SemaphoreIdentityPCDTypeName
+          ) {
+            return false;
+          }
+
+          // If the user has an email PCD, don't import another
+          if (userHasEmailPCD && pcd.type === EmailPCDTypeName) {
+            return false;
+          }
+
+          // If a PCD with this ID exists already, don't import it
+          if (pcdCollection.hasPCDWithId(pcd.id)) {
+            return false;
+          }
+
+          // Otherwise, do import it
+          return true;
+        };
+
+        const pcdsToMerge = importedCollection.getAll().filter(preImportFilter);
+        const pcdsToMergeIds = new Set(pcdsToMerge.map((pcd) => pcd.id));
+
+        // Because this hook can be called multiple times, we should only
```

Comment:
How does the UI flow allow this to be called multiple times?  Will it allow the user to actually pick multiple different files, or is it just re-renders with the same file?  If it's only re-renders can the content ever change based on the file input or only due to changes to the user's PCDCollection?

If these can be different files, then there would be an edge case if those files contain PCDs with the same IDs but different contents, where this block would think nothing had changed because it only compares IDs.  If you think that's a possibility, you might filter it out by checking a hash of the file to see if the input has changed.

Reply:
Based on my reading of the UI code below, I think the only way to pick multiple files is if the first one is invalid, which never reaches this line.

Reply:
Oh, good points. Generating a hash seems excessive, but then deep comparison in JavaScript is hard.